File: src/lib/ai.ts
import i18n from "@/i18n/config";
import { invoke } from "@tauri-apps/api/core";
import { listen } from "@tauri-apps/api/event";
import { nanoid } from "nanoid";

export interface AiMessage {
  role: string;
  content: string;
}

export interface AiProviderConfig {
  id: string;
  name: string;
  type: "preset" | "custom";
  provider?: "deepseek" | "alibaba";
  baseUrl: string;
  apiKey: string;
  model: string;
}

export interface AiCompleteRequest {
  providerConfigId: string;
  messages: AiMessage[];
  stream?: boolean;
  requestId?: string;
  temperature?: number;
}

export interface AiCompleteResponse {
  content: string;
}

interface AiStreamChunkEvent {
  requestId: string;
  content: string;
}

interface AiStreamDoneEvent {
  requestId: string;
  content: string;
}

const isTauriRuntime =
  typeof window !== "undefined" && ("__TAURI_INTERNALS__" in window || "__TAURI__" in window);

async function resolveProviderConfig(providerConfigId: string): Promise<AiProviderConfig> {
  if (isTauriRuntime) {
    const config = await invoke<AiProviderConfig | null>("get_ai_config", {
      id: providerConfigId,
    });
    if (!config) {
      throw new Error(
        i18n.t("common.aiCallFailed", {
          defaultValue: "AI configuration not found",
        }),
      );
    }
    return config;
  }

  const { useQuizStore } = await import("@/store/quizStore");
  const config = useQuizStore
    .getState()
    .settings.aiConfigs.find((item) => item.id === providerConfigId);
  if (!config) {
    throw new Error(
      i18n.t("common.aiCallFailed", {
        defaultValue: "AI configuration not found",
      }),
    );
  }
  return config;
}

function buildChatCompletionsUrl(baseUrl: string): string {
  if (baseUrl.endsWith("/chat/completions")) {
    return baseUrl;
  }

  return `${baseUrl.replace(/\/+$/, "")}/chat/completions`;
}

async function callDirectAI(
  config: AiProviderConfig,
  request: AiCompleteRequest,
  onChunk?: (chunk: string) => void,
): Promise<string> {
  const apiURL = buildChatCompletionsUrl(config.baseUrl);
  const response = await fetch(apiURL, {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
      Authorization: `Bearer ${config.apiKey}`,
    },
    body: JSON.stringify({
      model: config.model,
      messages: request.messages,
      stream: request.stream ?? false,
      temperature: request.temperature ?? 0.7,
    }),
  });

  if (!response.ok) {
    const errorData = await response.json().catch(() => ({}));
    throw new Error(
      errorData.error?.message ||
        i18n.t("common.apiFailed", {
          status: response.status,
          defaultValue: `API request failed: ${response.status}`,
        }),
    );
  }

  if (!request.stream) {
    const data = await response.json();
    return data.choices?.[0]?.message?.content || "";
  }

  const reader = response.body?.getReader();
  if (!reader) {
    throw new Error(
      i18n.t("common.streamReadFailed", {
        defaultValue: "Unable to read response stream",
      }),
    );
  }

  const decoder = new TextDecoder();
  let buffer = "";
  let fullText = "";

  while (true) {
    const { done, value } = await reader.read();
    if (done) break;

    buffer += decoder.decode(value, { stream: true });

    while (true) {
      const lineBreak = buffer.indexOf("\n");
      if (lineBreak === -1) break;

      const line = buffer.slice(0, lineBreak).trim();
      buffer = buffer.slice(lineBreak + 1);

      if (!line.startsWith("data:")) continue;

      const data = line.replace(/^data:\s*/, "");
      if (data === "[DONE]") continue;

      try {
        const parsed = JSON.parse(data);
        const content = parsed.choices?.[0]?.delta?.content || "";
        if (content) {
          fullText += content;
          onChunk?.(content);
        }
      } catch {
        // Ignore malformed partial chunks.
      }
    }
  }

  return fullText;
}

export async function callAI(providerConfigId: string, messages: AiMessage[]): Promise<string> {
  const request: AiCompleteRequest = {
    providerConfigId,
    messages,
    stream: false,
  };

  if (isTauriRuntime) {
    const response = await invoke<AiCompleteResponse>("ai_complete", {
      request,
    });
    return response.content;
  }

  const config = await resolveProviderConfig(providerConfigId);
  return callDirectAI(config, request);
}

export async function callAIStream(
  providerConfigId: string,
  messages: AiMessage[],
  onChunk: (chunk: string) => void,
): Promise<string> {
  const requestId = nanoid();
  const request: AiCompleteRequest = {
    providerConfigId,
    messages,
    stream: true,
    requestId,
  };

  if (isTauriRuntime) {
    let fullText = "";

    const unlistenChunk = await listen<AiStreamChunkEvent>("ai-stream:chunk", (event) => {
      if (event.payload.requestId !== requestId) return;
      fullText += event.payload.content;
      onChunk(event.payload.content);
    });

    const unlistenDone = await listen<AiStreamDoneEvent>("ai-stream:done", (event) => {
      if (event.payload.requestId !== requestId) return;
      fullText = event.payload.content;
    });

    try {
      const response = await invoke<AiCompleteResponse>("ai_complete", {
        request,
      });
      return response.content || fullText;
    } finally {
      unlistenChunk();
      unlistenDone();
    }
  }

  const config = await resolveProviderConfig(providerConfigId);
  return callDirectAI(config, request, onChunk);
}

File: src-tauri/src/ai.rs
use rusqlite::{params, Connection, OptionalExtension};
use serde::{Deserialize, Serialize};
use std::{
    path::Path,
    time::{SystemTime, UNIX_EPOCH},
};
use tauri::{Emitter, State, Window};

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct AiMessage {
    pub role: String,
    pub content: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct AiProviderConfig {
    pub id: String,
    pub name: String,
    #[serde(rename = "type")]
    pub config_type: String,
    pub provider: Option<String>,
    pub base_url: String,
    pub api_key: String,
    pub model: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct AiCompleteRequest {
    pub provider_config_id: String,
    pub messages: Vec<AiMessage>,
    pub stream: bool,
    pub request_id: Option<String>,
    pub temperature: Option<f32>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct AiCompleteResponse {
    pub content: String,
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
struct AiStreamChunkEvent {
    request_id: String,
    content: String,
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
struct AiStreamDoneEvent {
    request_id: String,
    content: String,
}

fn now_millis() -> i64 {
    SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .unwrap_or_default()
        .as_millis() as i64
}

fn open_connection(db_path: &Path) -> Result<Connection, String> {
    Connection::open(db_path).map_err(|error| error.to_string())
}

pub fn initialize_database(db_path: &Path) -> Result<(), String> {
    let conn = open_connection(db_path)?;
    conn.execute_batch(
        r#"
        CREATE TABLE IF NOT EXISTS ai_configs (
            id TEXT PRIMARY KEY,
            name TEXT NOT NULL,
            config_type TEXT NOT NULL,
            provider TEXT,
            base_url TEXT NOT NULL,
            api_key TEXT NOT NULL,
            model TEXT NOT NULL,
            created_at INTEGER NOT NULL,
            updated_at INTEGER NOT NULL
        );
        "#,
    )
    .map_err(|error| error.to_string())?;

    Ok(())
}

fn read_config(conn: &Connection, id: &str) -> Result<Option<AiProviderConfig>, String> {
    conn.query_row(
        r#"
        SELECT id, name, config_type, provider, base_url, api_key, model
        FROM ai_configs
        WHERE id = ?1
        "#,
        [id],
        |row| {
            Ok(AiProviderConfig {
                id: row.get(0)?,
                name: row.get(1)?,
                config_type: row.get(2)?,
                provider: row.get(3)?,
                base_url: row.get(4)?,
                api_key: row.get(5)?,
                model: row.get(6)?,
            })
        },
    )
    .optional()
    .map_err(|error| error.to_string())
}

fn save_config(conn: &Connection, config: &AiProviderConfig) -> Result<(), String> {
    let timestamp = now_millis();
    let existing = read_config(conn, &config.id)?;

    if existing.is_some() {
        conn.execute(
            r#"
            UPDATE ai_configs
            SET name = ?2,
                config_type = ?3,
                provider = ?4,
                base_url = ?5,
                api_key = ?6,
                model = ?7,
                updated_at = ?8
            WHERE id = ?1
            "#,
            params![
                config.id,
                config.name,
                config.config_type,
                config.provider,
                config.base_url,
                config.api_key,
                config.model,
                timestamp
            ],
        )
        .map_err(|error| error.to_string())?;
    } else {
        conn.execute(
            r#"
            INSERT INTO ai_configs (
                id, name, config_type, provider, base_url, api_key, model, created_at, updated_at
            )
            VALUES (?1, ?2, ?3, ?4, ?5, ?6, ?7, ?8, ?9)
            "#,
            params![
                config.id,
                config.name,
                config.config_type,
                config.provider,
                config.base_url,
                config.api_key,
                config.model,
                timestamp,
                timestamp
            ],
        )
        .map_err(|error| error.to_string())?;
    }

    Ok(())
}

fn delete_config(conn: &Connection, id: &str) -> Result<(), String> {
    conn.execute("DELETE FROM ai_configs WHERE id = ?1", [id])
        .map_err(|error| error.to_string())?;
    Ok(())
}

fn normalize_base_url(base_url: &str) -> String {
    if base_url.ends_with("/chat/completions") {
        base_url.to_string()
    } else {
        format!("{}/chat/completions", base_url.trim_end_matches('/'))
    }
}

async fn fetch_response_text(response: reqwest::Response) -> Result<String, String> {
    response.text().await.map_err(|error| error.to_string())
}

async fn request_ai_completion(
    config: AiProviderConfig,
    request: AiCompleteRequest,
    window: Option<Window>,
) -> Result<AiCompleteResponse, String> {
    if config.base_url.trim().is_empty() {
        return Err("AI base URL is required".to_string());
    }
    if config.api_key.trim().is_empty() {
        return Err("AI API key is required".to_string());
    }
    if config.model.trim().is_empty() {
        return Err("AI model is required".to_string());
    }

    let client = reqwest::Client::new();
    let response = client
        .post(normalize_base_url(&config.base_url))
        .bearer_auth(config.api_key)
        .json(&serde_json::json!({
            "model": config.model,
            "messages": request.messages,
            "temperature": request.temperature.unwrap_or(0.7),
            "stream": request.stream,
        }))
        .send()
        .await
        .map_err(|error| error.to_string())?;

    if !response.status().is_success() {
        let payload = fetch_response_text(response).await?;
        if let Ok(value) = serde_json::from_str::<serde_json::Value>(&payload) {
            if let Some(message) = value
                .pointer("/error/message")
                .and_then(|value| value.as_str())
            {
                return Err(message.to_string());
            }
        }
        return Err(payload);
    }

    if request.stream {
        let request_id = request
            .request_id
            .unwrap_or_else(|| format!("ai-{}", now_millis()));
        let mut buffer = String::new();
        let mut full_text = String::new();

        let mut response = response;
        while let Some(chunk) = response.chunk().await.map_err(|error| error.to_string())? {
            buffer.push_str(&String::from_utf8_lossy(&chunk));

            while let Some(split_index) = buffer.find('\n') {
                let line = buffer[..split_index].trim().to_string();
                buffer.drain(..split_index + 1);

                if !line.starts_with("data:") {
                    continue;
                }

                let data = line.trim_start_matches("data:").trim();
                if data == "[DONE]" {
                    if let Some(window) = &window {
                        let _ = window.emit(
                            "ai-stream:done",
                            AiStreamDoneEvent {
                                request_id: request_id.clone(),
                                content: full_text.clone(),
                            },
                        );
                    }
                    return Ok(AiCompleteResponse { content: full_text });
                }

                if let Ok(value) = serde_json::from_str::<serde_json::Value>(data) {
                    if let Some(content) = value
                        .pointer("/choices/0/delta/content")
                        .and_then(|value| value.as_str())
                    {
                        if !content.is_empty() {
                            full_text.push_str(content);
                            if let Some(window) = &window {
                                let _ = window.emit(
                                    "ai-stream:chunk",
                                    AiStreamChunkEvent {
                                        request_id: request_id.clone(),
                                        content: content.to_string(),
                                    },
                                );
                            }
                        }
                    }
                }
            }
        }

        if let Some(window) = &window {
            let _ = window.emit(
                "ai-stream:done",
                AiStreamDoneEvent {
                    request_id,
                    content: full_text.clone(),
                },
            );
        }

        return Ok(AiCompleteResponse { content: full_text });
    }

    let payload = response
        .json::<serde_json::Value>()
        .await
        .map_err(|error| error.to_string())?;

    let content = payload
        .pointer("/choices/0/message/content")
        .and_then(|value| value.as_str())
        .ok_or_else(|| "AI response did not include message content".to_string())?;

    Ok(AiCompleteResponse {
        content: content.to_string(),
    })
}

#[tauri::command]
pub async fn save_ai_config(
    state: State<'_, crate::AppState>,
    config: AiProviderConfig,
) -> Result<(), String> {
    let conn = open_connection(&state.db_path)?;
    save_config(&conn, &config)
}

#[tauri::command]
pub async fn delete_ai_config(state: State<'_, crate::AppState>, id: String) -> Result<(), String> {
    let conn = open_connection(&state.db_path)?;
    delete_config(&conn, &id)
}

#[tauri::command]
pub async fn get_ai_config(
    state: State<'_, crate::AppState>,
    id: String,
) -> Result<Option<AiProviderConfig>, String> {
    let conn = open_connection(&state.db_path)?;
    read_config(&conn, &id)
}

#[tauri::command]
pub async fn list_ai_configs(
    state: State<'_, crate::AppState>,
) -> Result<Vec<AiProviderConfig>, String> {
    let conn = open_connection(&state.db_path)?;
    let mut stmt = conn
        .prepare(
            r#"
            SELECT id, name, config_type, provider, base_url, api_key, model
            FROM ai_configs
            ORDER BY updated_at DESC
            "#,
        )
        .map_err(|error| error.to_string())?;

    let rows = stmt
        .query_map([], |row| {
            Ok(AiProviderConfig {
                id: row.get(0)?,
                name: row.get(1)?,
                config_type: row.get(2)?,
                provider: row.get(3)?,
                base_url: row.get(4)?,
                api_key: row.get(5)?,
                model: row.get(6)?,
            })
        })
        .map_err(|error| error.to_string())?;

    let mut configs = Vec::new();
    for row in rows {
        configs.push(row.map_err(|error| error.to_string())?);
    }
    Ok(configs)
}

#[tauri::command]
pub async fn ai_complete(
    state: State<'_, crate::AppState>,
    window: Window,
    request: AiCompleteRequest,
) -> Result<AiCompleteResponse, String> {
    let conn = open_connection(&state.db_path)?;
    let config = read_config(&conn, &request.provider_config_id)?
        .ok_or_else(|| "AI configuration not found".to_string())?;
    request_ai_completion(config, request, Some(window)).await
}
